carma-messenger-core/cpp_message/src/BSM_Message.cpp

```diff
@@ -0,0 +1,239 @@
+/*
+ * Copyright (C) 2020 LEIDOS.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+/**
+ * CPP File containing BSM Message method implementations
+ */
+
+#include "BSM_Message.h"
+
+namespace cpp_message
+{
+    boost::optional<j2735_msgs::BSM> BSM_Message::decode_bsm_message(std::vector<uint8_t>& binary_array){
+        
+        j2735_msgs::BSM output;
+        //decode results - stored in binary_array
+        asn_dec_rval_t rval;
+        MessageFrame_t* message = nullptr;
+        
+        //copy from vector to array         
+        size_t len=binary_array.size();    
+        
+        uint8_t buf[len];             
+        std::copy(binary_array.begin(),binary_array.end(),buf);
+        //use asn1c lib to decode
+        
+        rval=uper_decode(0, &asn_DEF_MessageFrame,(void **) &message, buf, len, 0, 0);
+         
+        //if decode success
+        if(rval.code==RC_OK)
+        {
+            BSMcoreData_t core_data_msg = message->value.choice.BasicSafetyMessage.coreData;
+            output.core_data.msg_count = core_data_msg.msgCnt; 
+            auto id_len = core_data_msg.id.size;
+            for(auto i = 0; i < id_len; i++)
+            {
+                output.core_data.id.push_back(core_data_msg.id.buf[i]);
+            }
+            output.core_data.sec_mark = core_data_msg.secMark;
+            output.core_data.latitude = core_data_msg.lat;
+            output.core_data.longitude = core_data_msg.Long; 
+            output.core_data.elev = core_data_msg.elev;
+            output.core_data.accuracy.orientation = core_data_msg.accuracy.orientation;
+            output.core_data.accuracy.semiMajor = core_data_msg.accuracy.semiMajor;
+            output.core_data.accuracy.semiMinor = core_data_msg.accuracy.semiMinor;
+            output.core_data.transmission.transmission_state = core_data_msg.transmission;
+            output.core_data.speed = core_data_msg.speed;
+            output.core_data.heading = core_data_msg.heading;
+            output.core_data.angle = core_data_msg.angle;
+            output.core_data.accelSet.lateral = core_data_msg.accelSet.lat;
+            output.core_data.accelSet.longitudinal =core_data_msg.accelSet.Long;
+            output.core_data.accelSet.vert = core_data_msg.accelSet.vert;
+            output.core_data.accelSet.yaw_rate = core_data_msg.accelSet.yaw;
+            // brake_applied_status decoding
```

Comment:
Two sets of comments here, do they imply the same information?

Reply:
this may be referring to the last comment but could you clarify?

Reply:
Your update resolves my doubt here